bsp/src/mill/bsp/ModuleUtils.scala

```diff
@@ -112,15 +111,24 @@ object ModuleUtils {
     target
   }
 
-  def getMillBuildClasspath(evaluator: Evaluator, source: Boolean): Seq[String] = {
-    val all = Try(evaluator.rootModule.getClass.getClassLoader.asInstanceOf[SpecialClassLoader]).fold(
+  def getMillBuildClasspath(evaluator: Evaluator, sources: Boolean): Seq[String] = {
+    val classpath = Try(evaluator.rootModule.getClass.getClassLoader.asInstanceOf[SpecialClassLoader]).fold(
       _ => Seq.empty,
       _.allJars
     )
-    val filtered =
-      if (source) all.filter(url => isSourceJar(url))
+
+    val millJars = resolveDependencies(
+      Resolve.defaultRepositories,
+      depToDependency(_, BuildInfo.scalaVersion),
+      BuildInfo.millEmbeddedModules.map(module => ivy"com.lihaoyi::mill-$module:${BuildInfo.millVersion}"),
```

Comment:
I think using full ivy artifact is the way to go (using `module.publishSelfDependency()`  when generating the `BuildInfo`). Less assumptions about the deps.

Reply:
Check out the new commit.